tuxedo-parachain-core/register_validate_block/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+//! This macro is copied from cumulus-pallet-parachain-system-proc-macro crate
+//! and modified slightly to fit Tuxedo's needs.
+
+use proc_macro2::Span;
+use proc_macro_crate::{crate_name, FoundCrate};
+use syn::{
+    parse::{Parse, ParseStream},
+    Error, Ident, Token,
+};
+
+/// Provides an identifier that is a safe way to refer to the crate tuxedo_core within the macro
+fn crate_() -> Result<Ident, Error> {
+    match crate_name("tuxedo-parachain-core") {
+        Ok(FoundCrate::Itself) => Ok(syn::Ident::new("tuxedo_parachain_core", Span::call_site())),
+        Ok(FoundCrate::Name(name)) => Ok(Ident::new(&name, Span::call_site())),
+        Err(e) => Err(Error::new(Span::call_site(), e)),
+    }
+}
+
+struct RegisterValidateBlockInput {
+    pub block: Ident,
+    _comma1: Token![,],
+    pub verifier: Ident,
+    _comma2: Token![,],
+    pub constraint_checker: Ident,
+}
+
+impl Parse for RegisterValidateBlockInput {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed = Self {
+            block: input.parse()?,
+            _comma1: input.parse()?,
+            verifier: input.parse()?,
+            _comma2: input.parse()?,
+            constraint_checker: input.parse()?,
+        };
+
+        if !input.is_empty() {
+            return Err(Error::new(
+                input.span(),
+                "Expected exactly three parameters: Block, Verifier, ConstraintChecker.",
+            ));
+        }
+
+        Ok(parsed)
+    }
+}
+
+#[proc_macro]
+pub fn register_validate_block(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    // Extract the paths to the parts from the runtime developer's input
+    // I will likely need to revise or simplify the fields that are passed in.
+    // I hope to only use the exposed runtime APIs here, not some custom trait impls. (if possible)
+    let input: RegisterValidateBlockInput = match syn::parse(input) {
+        Ok(t) => t,
+        Err(e) => return e.into_compile_error().into(),
+    };
+
+    let block = input.block.clone();
+    let verifier = input.verifier.clone();
+    let constraint_checker = input.constraint_checker.clone();
+
+    // A way to refer to the tuxedo_parachain_core crate from within the macro.
+    let crate_ = match crate_() {
+        Ok(c) => c,
+        Err(e) => return e.into_compile_error().into(),
+    };
+
+    //TODO We need to check inherents. At least the timestamp one, and maybe also the parachain one?
```

Comment:
Is this still relevant?

Reply:
Yes, and the followup issue is linked below. In terms of the grant, this part does not need to be delivered until milestone 2.